### PR TITLE
Refactor Visit Date and Time Format

### PR DESCRIPTION
**Pull request by EGiataganas**
We found that  this app has many different date formats such as "25 October 2015",
"Sunday 25 October 2015", "30/10/2015", and both 12 and 24-hour times.
Time periods are shown as "1:30pm to 3:00pm", "1:30pm, 1 hr 30 mins",
and "13:45-16:00".

According to GDS design pattern we will use the format "Sunday 25 October"
for Visit and "25 October 2015" for the Date of Birth. For time slots we will use
"1:30pm for 1h".

### CODE REVIEW

**Comment by EGiataganas** [comment]
@dan-palmer Thanks mentioning to use `describe` for test! :+1: Yeah I `add_template_helper(DateHelper)` because is required for email subject.

File: spec/helpers/date_helper_spec.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  describe '.display_start_time' do
+    it 'formats a start time from a 24hr time string' do
+      expect(helper.display_start_time('0945')).to eq('9:45am')
+    end
+  end
+
+  describe '.display_slot_and_duration' do
+    it 'displays a slot and duration' do
+      expect(helper.display_slot_and_duration('0945-1145')).to eq('9:45am for 2 hrs')
+      expect(helper.display_slot_and_duration('1430-1600')).to eq('2:30pm for 1 hr 30 mins')
```
Comment (SteveMarshall) [inline spec/helpers/date_helper_spec.rb]: Do we know how these abbreviations are handled (particularly by screen readers)? I suspect not very well…

Comment (SteveMarshall) [inline spec/helpers/date_helper_spec.rb]: Oh, I've just tested in VoiceOver on my Mac and apparently it works just fine there!

File: spec/helpers/date_helper_spec.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    it 'formats a time in 24 clock from a string' do
+      expect(helper.format_time_str_24('0900')).to eq('09:00')
+      expect(helper.format_time_str_24('1200')).to eq('12:00')
+      expect(helper.format_time_str_24('1545')).to eq('15:45')
+    end
+  end
+
+  describe '.display_start_time' do
+    it 'formats a start time from a 24hr time string' do
+      expect(helper.display_start_time('0945')).to eq('9:45am')
```
Comment (dan-palmer) [inline spec/helpers/date_helper_spec.rb]: Looks as if this method performs the same task as `format_time_str` does.

**Comment by dan-palmer** [comment]
> @dan-palmer Thanks mentioning to use describe for test! :+1: Yeah I add_template_helper(DateHelper) because is required for email subject.

@EGiataganas Its included for that purpose, the `add_template_helper ` makes it available when rendering the mails view.

File: app/helpers/date_helper.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
+module DateHelper
+  def format_date_of_birth(date)
+    I18n.l(date_from_string_or_date(date), format: :date_of_birth)
+  end
+
+  def format_date_of_visit(date)
+    I18n.l(date_from_string_or_date(date), format: :date_of_visit)
+  end
+
+  def format_time_str(time)
```
Comment (dan-palmer) [inline app/helpers/date_helper.rb]: Seems as we have a method called `format_time_str_24 ` would it make sense to call this `format_time_str_12`? Current method name is a little on the ambiguous side.

Comment (SteveMarshall) [inline app/helpers/date_helper.rb]: I might be inclined toward `format_time_12hr` and `format_time_24hr`?

File: spec/helpers/date_helper_spec.rb (hunk 2 above)
Comment (dan-palmer) [inline spec/helpers/date_helper_spec.rb]: Looked at the method and it seems as though this test can be improved. It should be passed a string time range e.g 
`'0945-1015'` & is expected to return the beginning of the range in its 12 hour format.

File: spec/helpers/date_helper_spec.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    it 'formats a day from a string' do
+      expect(helper.format_date_of_visit('2014-07-24')).to eq('Thursday 24 July')
+    end
+    it 'formats a day from a date' do
+      expect(helper.format_date_of_visit(Date.parse('2014-07-24'))).to eq('Thursday 24 July')
+    end
+  end
+
+  describe '.format_time_str' do
+    it 'formats a time from a string' do
```
Comment (dan-palmer) [inline spec/helpers/date_helper_spec.rb]: Something like: 
```ruby
it 'formats the time from 24 hour formatted string to a 12 hour formatted string' do

File: app/helpers/date_helper.rb (hunk 3 above)
Comment (EGiataganas) [inline app/helpers/date_helper.rb]: @dan-palmer and @SteveMarshall Thank you, I will make that change :+1:

Comment (dan-palmer) [inline app/helpers/date_helper.rb]: :+1:

File: spec/helpers/date_helper_spec.rb (hunk 2 above)
Comment (EGiataganas) [inline spec/helpers/date_helper_spec.rb]: @dan-palmer yeap, you re right! I have to changed it to `0945-1145`

File: app/helpers/date_helper.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    format_time_12hr(times.split('-')[0])
+  end
+
+  def display_slot_and_duration(times, glue = ' for ')
+    from, to = times.split('-')
+    [
+      display_start_time(times),
+      (time_from_string(to) - time_from_string(from)).duration
+    ].join(glue)
+  end
```
Comment (threedaymonk) [inline app/helpers/date_helper.rb]: I think it might be more consistent for these `display_*` methods to be `format_*` like the others.

Comment (EGiataganas) [inline app/helpers/date_helper.rb]: @threedaymonk No problem!!

File: app/views/slots/edit.html.haml [hunk 6]
```diff
@@ -87,7 +87,7 @@
               - @schedule.available_visitation_dates.each do |date|
                 - slots_for_day(date).each do |slot|
                   - value = [date.strftime('%Y-%m-%d'), slot.join('-')].join('-')
-                  %option{ value: value, selected: ('selected' if value == current_slots[index]) }= date.strftime('%A, %e %B - ') + Time.strptime(slot[0].to_s, '%H%M').strftime('%l:%M%P')
+                  %option{ value: value, selected: ('selected' if value == current_slots[index]) }= format_date_of_visit(date) + Time.strptime(slot[0].to_s, '%H%M').strftime('%l:%M%P')
```
Comment (tatyree) [inline app/views/slots/edit.html.haml]: We should consider moving some of these long-lined nested loops into helpers.

File: app/views/visitor_mailer/booking_receipt_email.text.erb [hunk 7]
```diff
@@ -19,7 +19,7 @@ You have requested a visit to <%= @visit.prison_name %> for <%= @visit.visitors.
 
 <% @visit.slots.each_with_index do |slot, index| %>
 <% if slot.date.present? %>
-<% if index == 0 %>First choice: <% else %>Alternative <%= index %>:<% end %> <%= Date.parse(slot.date).strftime("%A %e %B") %>, <%= display_time_slot(slot.times) %>
+<% if index == 0 %>First choice: <% else %>Alternative <%= index %>:<% end %> <%= format_date_of_visit(slot.date) %>, <%= format_slot_and_duration(slot.times) %>
```
Comment (tatyree) [inline app/views/visitor_mailer/booking_receipt_email.text.erb]: I'm uncomfortable with if/then/else blocks on a single line in a view.

File: spec/helpers/date_helper_spec.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    end
+  end
+
+  describe '.format_slot_and_duration' do
+    it 'displays a slot and duration' do
+      expect(helper.format_slot_and_duration('0945-1145')).to eq('9:45am for 2 hrs')
+      expect(helper.format_slot_and_duration('1430-1600')).to eq('2:30pm for 1 hr 30 mins')
+    end
+  end
+end
```
Comment (tatyree) [inline spec/helpers/date_helper_spec.rb]: :+1:

File: config/locales/en.yml [hunk 9]
```diff
@@ -1,12 +1,16 @@
...
-      long: "%A %-d %B %Y"
+      date_of_birth: "%-d %B %Y" # 24 July 2014
+      date_of_visit: "%A %-d %B" # Thursday 24 July
     fields:
       day: Day
       month: Month
       year: Year
+  time:
+    formats:
+      twelve_hour: "%-l:%M%P"
```
Comment (dan-palmer) [inline config/locales/en.yml]: Fan of the comment on the date formats, perhaps do the same here.

**Comment by EGiataganas** [comment]
Tested on preprod

File: app/helpers/visit_helper.rb [hunk 10]
```diff
@@ -61,7 +61,7 @@ def slots_for_day(day)
 
   def when_to_expect_reply
     schedule = PrisonSchedule.new(prison)
-    format_day(schedule.confirmation_email_date)
+    format_date_of_visit(schedule.confirmation_email_date)
```
Comment (tatyree) [inline app/helpers/visit_helper.rb]: Shouldn't this be format_date_of_reply?

File: app/helpers/date_helper.rb [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    [
+      format_date_of_visit(slot.date),
+      format_slot_and_duration(slot.times)
+    ].join(' ')
+  end
+
+  def date_and_times_of_slot(slot)
+    [
+      format_date_of_visit(slot.date),
+      stat_and_end_time(slot.times)
```
Comment (threedaymonk) [inline app/helpers/date_helper.rb]: There's a typo: should be `start_and_end_time`.

File: app/helpers/date_helper.rb [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+  def format_slot_and_duration(times, glue = ' for ')
+    from, to = times.split('-')
+    [
+      format_start_time(times),
+      (time_from_string(to) - time_from_string(from)).duration
+    ].join(glue)
+  end
+
+  def stat_and_end_time(times)
```
Comment (threedaymonk) [inline app/helpers/date_helper.rb]: The same typo (at least it's consistent! :smile:)

File: spec/helpers/date_helper_spec.rb [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    it 'formats a day from a string' do
+      expect(helper.format_date('2014-07-24')).to eq('Thursday 24 July')
+    end
+    it 'formats a day from a date' do
+      expect(helper.format_date(Date.parse('2014-07-24'))).to eq('Thursday 24 July')
+    end
+  end
+
+  describe '.format_time_12hr' do
+    it 'formats a time from a string' do
```
Comment (dan-palmer) [inline spec/helpers/date_helper_spec.rb]: Perhaps: `it 'formats a 24hr time string into a 12hr clock format'`

File: spec/helpers/date_helper_spec.rb [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+      expect(helper.format_time_12hr('0900')).to eq('9:00am')
+      expect(helper.format_time_12hr('1200')).to eq('12:00pm')
+      expect(helper.format_time_12hr('1545')).to eq('3:45pm')
+    end
+  end
+
+  describe '.format_time_24hr' do
+    it 'formats a time in 24 clock from a string' do
```
Comment (dan-palmer) [inline spec/helpers/date_helper_spec.rb]: Perhaps: `it 'formats a 24hr time string with a separator between hours & minutes'`

File: spec/helpers/date_helper_spec.rb (hunk 13 above)
Comment (EGiataganas) [inline spec/helpers/date_helper_spec.rb]: @dan-palmer :+1:

File: spec/helpers/date_helper_spec.rb (hunk 14 above)
Comment (EGiataganas) [inline spec/helpers/date_helper_spec.rb]: @dan-palmer :+1:

File: vendor/assets/moj.slot-picker/dist/javascripts/moj.slot-picker.js [hunk 15]
```diff
@@ -324,10 +324,10 @@
           time = label.find('.SlotPicker-time').text(),
           duration = label.find('.SlotPicker-duration').text(),
           $slot = this.$choice.eq(index);
-      
+
       $slot.addClass('is-chosen');
       $slot.find('.SlotPicker-date').text(day);
-      $slot.find('.SlotPicker-time').text(time + ', ' + duration);
+      $slot.find('.SlotPicker-time').text(time + ' for ' + duration);
```
Comment (dan-palmer) [inline vendor/assets/moj.slot-picker/dist/javascripts/moj.slot-picker.js]: Shouldn't this be made in the JS project.

File: app/helpers/date_helper.rb [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+  def time_from_string(obj)
+    Time.strptime(obj, '%H%M')
+  end
+
+  def format_start_time(times)
+    format_time_12hr(times.split('-').first)
+  end
+
+  def format_slot_and_duration(times, glue = ' for ')
+    from, to = times.split('-')
```
Comment (dan-palmer) [inline app/helpers/date_helper.rb]: `times.split('-')` is used a lot here to access the 'from' & 'to' times.

File: app/mailers/prison_mailer.rb [hunk 17]
```diff
@@ -108,7 +110,7 @@ def visitors_email_address
     visit.visitors.first.email
   end
 
-  def formatted_first_slot_date
-    I18n.l(Date.parse(@visit.slots.first.date), format: :long)
+  def first_date
+    @visit.slots.first.date
```
Comment (dan-palmer) [inline app/mailers/prison_mailer.rb]: Should probably just be able to ask a `Visit` what the first slot date is.

**Pull request merged**